File: src/update/update.ts
/**
 * 「检查新版本」的轻量实现，对齐 Go 版 internal/update。
 *
 * 不走 GitHub API（仅用 releases/latest 的 302 重定向抠版本号，无速率限制），结果缓存在
 * ~/.cc-mini/update-check.json，每 24h 才真去网络。显示永远读缓存（瞬时、不阻塞），过期时后台
 * 异步刷新——新版本「下次打开」才提示。离线/失败一律静默。只写工具自己的 ~/.cc-mini/（铁律）。
 */
import { existsSync, mkdirSync, readFileSync, renameSync, writeFileSync } from 'node:fs';
import { join } from 'node:path';

export const MODE_OFF = '';
export const MODE_NOTIFY = 'notify';

const LATEST_URL = 'https://github.com/becomeless/cc-x/releases/latest';
const CACHE_MAX_AGE_MS = 24 * 60 * 60 * 1000;
const HTTP_TIMEOUT_MS = 2000;
const CACHE_FILE = 'update-check.json';
const UPGRADE_CMD = 'npm i -g @cc-x/cc-x@latest';

interface Cache {
  checkedAt: number; // unix 秒
  latest: string;
}

const TAG_RE = /\/tag\/v?(\d+\.\d+\.\d+)/;

/** 读缓存：若已知有比 current 更新的版本，返回最新版本号；否则 undefined。不联网。 */
export function banner(storeDir: string, current: string): string | undefined {
  const c = readCache(storeDir);
  if (!c || !c.latest) return undefined;
  return isNewer(c.latest, current) ? c.latest : undefined;
}

/** 缓存过期（或不存在）时后台异步联网刷新一次；不阻塞调用方。 */
export function maybeRefresh(storeDir: string): void {
  const c = readCache(storeDir);
  if (c && Date.now() - c.checkedAt * 1000 < CACHE_MAX_AGE_MS) return; // 仍新鲜
  void refresh(storeDir); // fire-and-forget
}

async function refresh(storeDir: string): Promise<void> {
  const latest = await fetchLatest();
  if (!latest) return; // 失败静默；不动缓存
  writeCache(storeDir, { checkedAt: Math.floor(Date.now() / 1000), latest });
}

/** 当前版本（npm 版）的升级命令。 */
export function upgradeCommand(): string {
  return UPGRADE_CMD;
}

function cachePath(storeDir: string): string {
  return join(storeDir, CACHE_FILE);
}

function readCache(storeDir: string): Cache | undefined {
  try {
    const c = JSON.parse(readFileSync(cachePath(storeDir), 'utf-8')) as Cache;
    if (typeof c.checkedAt === 'number' && typeof c.latest === 'string') return c;
  } catch {
    /* 无缓存 / 损坏 -> 当作没有 */
  }
  return undefined;
}

/** 原子写（temp + rename），避免被进程退出打断写出半截文件。 */
function writeCache(storeDir: string, c: Cache): void {
  try {
    if (!existsSync(storeDir)) mkdirSync(storeDir, { recursive: true });
    const tmp = `${cachePath(storeDir)}.tmp`;
    writeFileSync(tmp, JSON.stringify(c), 'utf-8');
    renameSync(tmp, cachePath(storeDir));
  } catch {
    /* 写不了就算了 */
  }
}

/** 用 releases/latest 的 302 重定向抠最新版本号；redirect:'manual' = 不跟随 = 不走 GitHub API。 */
async function fetchLatest(): Promise<string | undefined> {
  try {
    const resp = await fetch(LATEST_URL, {
      redirect: 'manual',
      signal: AbortSignal.timeout(HTTP_TIMEOUT_MS),
    });
    const loc = resp.headers.get('location');
    if (!loc) return undefined;
    const m = TAG_RE.exec(loc);
    return m ? m[1] : undefined;
  } catch {
    return undefined;
  }
}

/** latest 是否严格新于 current（"a.b.c"，忽略前导 v 与后缀）。解析失败一律 false（不误报）。 */
export function isNewer(latest: string, current: string): boolean {
  const lp = parseSemver(latest);
  const cp = parseSemver(current);
  if (!lp || !cp) return false;
  const [la, lb, lc] = lp;
  const [ca, cb, cc] = cp;
  if (la !== ca) return la > ca;
  if (lb !== cb) return lb > cb;
  return lc > cc;
}

function parseSemver(s: string): [number, number, number] | undefined {
  let v = s.trim().replace(/^v/, '');
  const cut = v.search(/[-+]/);
  if (cut >= 0) v = v.slice(0, cut);
  const parts = v.split('.');
  if (parts.length !== 3) return undefined;
  const nums = parts.map((p) => Number.parseInt(p, 10));
  if (nums.some((n) => Number.isNaN(n))) return undefined;
  return [nums[0]!, nums[1]!, nums[2]!];
}

File: internal/update/update.go
// Package update 是「检查新版本」的轻量实现：不走 GitHub API（仅用 releases/latest 的 302 重定向
// 抠版本号，无速率限制），结果缓存在 ~/.cc-mini/update-check.json，每 24h 才真去网络。
//
// 设计：显示永远读缓存（瞬时、不阻塞），过期时后台异步刷新——所以新版本「下次打开」才提示，
// 与用户预期一致。离线/失败一律静默。只写工具自己的 ~/.cc-mini/，不碰任何 Claude Code 配置（铁律）。
//
// 网络请求通过 curl 子进程完成（exec.CommandContext），避免引入 net/http 拉入整个 TLS/加密栈。
// curl 在所有目标平台均可用：Windows 10 1803+ 内置 curl.exe；macOS/Linux 系统自带。
package update

import (
	"context"
	"encoding/json"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"runtime"
	"strconv"
	"strings"
	"time"
)

// Mode 是更新检查模式。空 = 关闭。
const (
	ModeOff    = ""
	ModeNotify = "notify"
)

const (
	latestURL    = "https://github.com/becomeless/cc-x/releases/latest"
	cacheMaxAge  = 24 * time.Hour
	fetchTimeout = 5 * time.Second
	cacheFile    = "update-check.json"
	winUpgrade   = "irm https://github.com/becomeless/cc-x/releases/latest/download/install.ps1 | iex"
	unixUpgrade  = "curl -fsSL https://github.com/becomeless/cc-x/releases/latest/download/install.sh | sh"
)

type cache struct {
	CheckedAt int64  `json:"checkedAt"`
	Latest    string `json:"latest"`
}

// tagRe 从 .../releases/tag/v0.4.3 抠出 0.4.3。
var tagRe = regexp.MustCompile(`/tag/v?(\d+\.\d+\.\d+)`)

// Banner 读缓存：若已知有比 current 更新的版本，返回 (最新版本号, true)。不联网。
func Banner(storeDir, current string) (string, bool) {
	c, err := readCache(storeDir)
	if err != nil || c.Latest == "" {
		return "", false
	}
	if isNewer(c.Latest, current) {
		return c.Latest, true
	}
	return "", false
}

// MaybeRefresh 缓存过期（或不存在）时后台异步联网刷新一次；不阻塞调用方。
func MaybeRefresh(storeDir string) {
	c, err := readCache(storeDir)
	if err == nil && time.Since(time.Unix(c.CheckedAt, 0)) < cacheMaxAge {
		return // 仍新鲜
	}
	go func() {
		latest := fetchLatest()
		if latest == "" {
			return // 失败静默；不动缓存
		}
		writeCache(storeDir, cache{CheckedAt: time.Now().Unix(), Latest: latest})
	}()
}

// UpgradeCommand 返回当前平台的升级命令（Go 原生版走安装器一行命令）。
func UpgradeCommand() string {
	if runtime.GOOS == "windows" {
		return winUpgrade
	}
	return unixUpgrade
}

func cachePath(storeDir string) string { return filepath.Join(storeDir, cacheFile) }

func readCache(storeDir string) (cache, error) {
	var c cache
	data, err := os.ReadFile(cachePath(storeDir))
	if err != nil {
		return c, err
	}
	if err := json.Unmarshal(data, &c); err != nil {
		return cache{}, err
	}
	return c, nil
}

// writeCache 原子写（temp + rename），避免后台 goroutine 被进程退出打断写出半截文件。
func writeCache(storeDir string, c cache) {
	if err := os.MkdirAll(storeDir, 0o755); err != nil {
		return
	}
	data, err := json.Marshal(c)
	if err != nil {
		return
	}
	tmp := cachePath(storeDir) + ".tmp"
	if err := os.WriteFile(tmp, data, 0o644); err != nil {
		return
	}
	_ = os.Rename(tmp, cachePath(storeDir))
}

// fetchLatest 用 curl 子进程获取 releases/latest 的 302 Location 并抠出版本号。
// --max-redirs 0 禁止跟随重定向；-w "%{redirect_url}" 直接输出目标 URL，无需解析响应头。
// curl 不可用或网络失败均静默返回 ""。
func fetchLatest() string {
	nullDev := "/dev/null"
	if runtime.GOOS == "windows" {
		nullDev = "NUL"
	}
	ctx, cancel := context.WithTimeout(context.Background(), fetchTimeout)
	defer cancel()
	out, err := exec.CommandContext(ctx, "curl", "-s", "--max-redirs", "0",
		"-o", nullDev, "-w", "%{redirect_url}", latestURL).Output()
	if err != nil || len(out) == 0 {
		return ""
	}
	m := tagRe.FindStringSubmatch(strings.TrimSpace(string(out)))
	if m == nil {
		return ""
	}
	return m[1]
}

// isNewer 报告 latest 是否严格新于 current（"a.b.c"，忽略前导 v 与后缀）。解析失败一律 false（不误报）。
func isNewer(latest, current string) bool {
	lp, ok1 := parseSemver(latest)
	cp, ok2 := parseSemver(current)
	if !ok1 || !ok2 {
		return false
	}
	for i := 0; i < 3; i++ {
		if lp[i] != cp[i] {
			return lp[i] > cp[i]
		}
	}
	return false
}

// parseSemver 解析 "v1.2.3" / "1.2.3-rc1" 的前三段数字。
func parseSemver(s string) ([3]int, bool) {
	var out [3]int
	s = strings.TrimPrefix(strings.TrimSpace(s), "v")
	// 砍掉 -rc1 / +build 之类后缀
	if i := strings.IndexAny(s, "-+"); i >= 0 {
		s = s[:i]
	}
	parts := strings.Split(s, ".")
	if len(parts) != 3 {
		return out, false
	}
	for i, p := range parts {
		n, err := strconv.Atoi(p)
		if err != nil {
			return out, false
		}
		out[i] = n
	}
	return out, true
}
